### ISSUE TITLE
abfs.ls('container/folder/') strips last '/'

### ISSUE THREAD

**Issue by mgsnuno** [dask/adlfs#140]
**What happened**:
`abfs.ls('container/folder/')` and `abfs.ls('container/folder')` produce the same result but should not. 

This causes problems when there are `folder`, `folder1`, `folder2` etc. It wont just list the contents of `folder`.

**Minimal Complete Verifiable Example**:
container structure
```
container
  folder
    ...
  folder1
    ...
  folder2
    ...
```
```python
abfs.ls('container/folder')  # ['container/folder/...', 'container/folder1/...', 'container/folder2/...']
abfs.ls('container/folder/')  # should be only ['container/folder/...'], but is same as above
```

**Anything else we need to know?**:

I think the culprit is this `strip`: https://github.com/dask/adlfs/blob/b43428914320e68962378f39a7183cfe01a2bcdd/adlfs/spec.py#L670

**Environment**:

- adlfs version: 0.5.7
- Python version: 3.8
- Operating System: linux
- Install method (conda, pip, source): conda/pip

**Comment by mgsnuno** [comment]
@hayesgb is it possible to remote that `target_path = path.strip("/")` without further implications downstream?

**Comment by mgsnuno** [comment]
@lostmygithubaccount can you chime in on this? thank you

**Comment by hayesgb** [comment]
@mgsnuno -- As written, setting that line to `target_path=path`, causes `fs.ls("container/folder/")` to return an empty list, rather than the the `['container/folder/']`

Additionally, I compared the result returned from fsspec.implementations.local.LocalFileSystem, and the results returned from abfs are consistent with the LocalFileSystem result.

What is your use case?

**Comment by mgsnuno** [comment]
My use case having a azure container with several parquet tables dumps, some tables share the initial name like I mentioned: users, users_1, users_2 etc. If everything was users_X it would be fine, but I really have a folder named users only, and it's not possible to list its contents alone.

**Comment by mgsnuno** [comment]
Adding some extra info: 
```
fs = fsspec.filesystem("abfs", **storage_options)
fs.glob('users/*')  # returns []
fs.glob('users_1/*')  # returns [users_1/...]
```

**Comment by ldacey** [comment]
I think I ran into this issue as well with pyarrow - I have some datasets with partitions that look like this:

```
/tgse-plus/
/tgse/
```

If I try to read data in the tgse partition, I run into an error that a file cannot be found in the tgse-plus folder even though I am not interested in that data (it is an entirely different sub domain).

**Comment by hayesgb** [comment]
I've pasted one of the tests that is in the current release and v0.3.3 below.  Can you propose a test that, if passed, would solve your issue?

```   
# root contains files and directories
    assert fs.ls("data/root") == [
        "data/root/a/",
        "data/root/b/",
        "data/root/c/",
        "data/root/rfile.txt",
    ]
    assert fs.ls("data/root/") == [
        "data/root/a/",
        "data/root/b/",
        "data/root/c/",
        "data/root/rfile.txt",
    ]
```

**Comment by mgsnuno** [comment]
build root structure such it contains:
```
"data/root/a/file.txt", "data/root/a1/file1.txt"
```
now the tests
```
assert fs.ls("data/root/a") == [
        "data/root/a/file.txt",
    ]
assert fs.ls("data/root/a/") == [
        "data/root/a/file.txt",
    ]
assert fs.ls("data/root/a1") == [
        "data/root/a1/file1.txt",
    ]
assert fs.ls("data/root/a1/") == [
        "data/root/a1/file.txt",
    ]
```

**Comment by hayesgb** [comment]
We currently have a root structure such that:
```
    container_client.upload_blob("top_file.txt", data)
    container_client.upload_blob("root/rfile.txt", data)
    container_client.upload_blob("root/a/file.txt", data)
    container_client.upload_blob("root/b/file.txt", data)
    container_client.upload_blob("root/c/file1.txt", data)
    container_client.upload_blob("root/c/file2.txt", data)
```

this test currently passes

```
    ## slashes are not not needed, but accepted
    assert fs.ls("data/root/a") == ["data/root/a/file.txt"]
    assert fs.ls("data/root/a/") == ["data/root/a/file.txt"]
    assert fs.ls("/data/root/a") == ["data/root/a/file.txt"]
    assert fs.ls("/data/root/a/") == ["data/root/a/file.txt"]
    assert fs.ls("data/root/b") == ["data/root/b/file.txt"]
    assert fs.ls("data/root/b/") == ["data/root/b/file.txt"]
```

I added the last 2 lines just to verify, but doesn't this reflect your test above?

**Comment by mgsnuno** [comment]
No it doesn't, please read carefully my input, what is needed to trigger this error is that there are more than 1 directory inside `data/root` that begins with the same name. In your example, there is only 1 `data/root/a`, it my example there is `data/root/a` and `data/root/a1`.

**Comment by lostmygithubaccount** [comment]
I ran into this working with the nyc taxi data from Azure open data, i.e.:

```python
from adlfs import AzureBlobFileSystem

color = "green"
container_name = "nyctlc"
storage_options = {"account_name": "azureopendatastorage"}

fs = AzureBlobFileSystem(**storage_options)

files = fs.ls(f"{container_name}/{color}/puYear=2016/puMonth=1/")
print(files)
```

this results in:

```python
['nyctlc/green/puYear=2016/puMonth=1', 'nyctlc/green/puYear=2016/puMonth=1/_SUCCESS', 'nyctlc/green/puYear=2016/puMonth=1/_committed_3144095708418575143', 'nyctlc/green/puYear=2016/puMonth=1/_committed_4753095944193949832', 'nyctlc/green/puYear=2016/puMonth=1/_committed_vacuum5568920582204786872', 'nyctlc/green/puYear=2016/puMonth=1/_started_4753095944193949832', 'nyctlc/green/puYear=2016/puMonth=1/part-00119-tid-4753095944193949832-fee7e113-666d-4114-9fcb-bcd3046479f3-2689-1.c000.snappy.parquet', 'nyctlc/green/puYear=2016/puMonth=10', 'nyctlc/green/puYear=2016/puMonth=10/_SUCCESS', 'nyctlc/green/puYear=2016/puMonth=10/_committed_3144095708418575143', 'nyctlc/green/puYear=2016/puMonth=10/_committed_4753095944193949832', 'nyctlc/green/puYear=2016/puMonth=10/_committed_vacuum4405553839532015818', 'nyctlc/green/puYear=2016/puMonth=10/_started_4753095944193949832', 'nyctlc/green/puYear=2016/puMonth=10/part-00090-tid-4753095944193949832-fee7e113-666d-4114-9fcb-bcd3046479f3-2660-1.c000.snappy.parquet', 'nyctlc/green/puYear=2016/puMonth=11', 'nyctlc/green/puYear=2016/puMonth=11/_SUCCESS', 'nyctlc/green/puYear=2016/puMonth=11/_committed_3144095708418575143', 'nyctlc/green/puYear=2016/puMonth=11/_committed_4753095944193949832', 'nyctlc/green/puYear=2016/puMonth=11/_committed_vacuum5178236545522498195', 'nyctlc/green/puYear=2016/puMonth=11/_started_4753095944193949832', 'nyctlc/green/puYear=2016/puMonth=11/part-00021-tid-4753095944193949832-fee7e113-666d-4114-9fcb-bcd3046479f3-2591-1.c000.snappy.parquet', 'nyctlc/green/puYear=2016/puMonth=12', 'nyctlc/green/puYear=2016/puMonth=12/_SUCCESS', 'nyctlc/green/puYear=2016/puMonth=12/_committed_3144095708418575143', 'nyctlc/green/puYear=2016/puMonth=12/_committed_4753095944193949832', 'nyctlc/green/puYear=2016/puMonth=12/_committed_vacuum6103530849850067575', 'nyctlc/green/puYear=2016/puMonth=12/_started_4753095944193949832', 'nyctlc/green/puYear=2016/puMonth=12/part-00116-tid-4753095944193949832-fee7e113-666d-4114-9fcb-bcd3046479f3-2686-1.c000.snappy.parquet']
`

which contains `puMonth=11` and `puMonth=12`

**Comment by hayesgb** [comment]
This was happening because the Azure SDK walk_blobs method takes a parameter called blob_prefix, which is not a filepath, but rather a string.  That string returns partial matches for blob names.  I've added @mgsnuno's test case on branch fixls, and it's passing.  It would be great to get some feedback before I merge into master.

**Comment by mgsnuno** [comment]
It is working without any problems now. Thanks a lot @hayesgb

**Comment by hayesgb** [comment]
Fixed with #147
